src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php

```diff
@@ -0,0 +1,155 @@
+
+            if (false === $doctrineEnvelope) {
+                $this->driverConnection->commit();
+
+                return null;
+            }
+
+            $doctrineEnvelope['headers'] = \json_decode($doctrineEnvelope['headers'], true);
+
+            $this->driverConnection->createQueryBuilder()
+                ->update($this->tableName)
+                ->set('status', ':delivered_status')
+                ->set('delivered_at', ':delivered_at')
+                ->where('id = :id')
+                ->setParameter('delivered_status', self::STATUS_DELIVERED)
+                ->setParameter('delivered_at', (new \DateTime('now'))->format('Y-m-d H:i:s'))
+                ->setParameter('id', $doctrineEnvelope['id'])
+                ->execute();
+
+            $this->driverConnection->commit();
+
+            return $doctrineEnvelope;
+        } catch (\Exception $e) {
+            $this->driverConnection->rollBack();
+
+            throw $e;
+        }
+    }
+
+    public function ack($id): bool
+    {
+        return $this->driverConnection->delete($this->tableName, array('id' => $id)) > 0;
+    }
+
+    public function nack($id): bool
+    {
+        return $this->driverConnection->update($this->tableName, array('status' => self::STATUS_PENDING), array('id' => $id)) > 0;
+    }
+
+    public function requeueMessages()
+    {
+        $redeliverLimit = new \DateTime('now');
+        $redeliverLimit->modify(sprintf('-%d second', $this->redeliverTimeout));
+        $this->driverConnection->createQueryBuilder()
+            ->update($this->tableName)
+            ->set('status', ':pending_status')
+            ->set('delivered_at', ':delivered_at')
+            ->where('status = :delivered_status')
+            ->andWhere('delivered_at < :redeliver_limit')
+            ->setParameter('pending_status', self::STATUS_PENDING)
+            ->setParameter('delivered_at', null)
+            ->setParameter('delivered_status', self::STATUS_DELIVERED)
+            ->setParameter('redeliver_limit', $redeliverLimit->format('Y-m-d H:i:s'))
+            ->execute();
+    }
+
+    private function setup(): void
+    {
+        $synchronizer = new SingleDatabaseSynchronizer($this->driverConnection);
+        $synchronizer->updateSchema($this->getSchema());
+    }
+
+    private function getSchema(): Schema
+    {
+        $schema = new Schema();
+        $table = $schema->createTable($this->tableName);
+        $table->addColumn('id', Type::BIGINT)
+            ->setAutoincrement(true)
+            ->setNotnull(true);
+        $table->addColumn('body', Type::TEXT)
+            ->setNotnull(true);
+        $table->addColumn('headers', Type::TEXT)
+            ->setNotnull(true);
+        $table->addColumn('status', Type::STRING)
+            ->setLength(32)
+            ->setDefault(self::STATUS_PENDING);
+        $table->addColumn('delivered_at', Type::DATETIME)
+            ->setNotnull(false);
+        $table->setPrimaryKey(array('id'));
```

Comment:
A few suggested changes:

* A `created_at` that's auto-populated
* Instead of `status`, what about an `available_at` set to a date? That would allow (in the future) for a delivery delay to be naturally added (as you could set things to a future date)
* Instead of `status`, what about also a `handling_at` set to a date - we would set this when we *start* handling the message. Ultimately, we would delete the row once the message has been fully handed.

I'm not sure if the two above fields are not taking into account any other statuses needed. I double-checked with Laravel's DB "transport", and they use two very similar columns to do all of the work needed.

Reply:
I also realize that this would not allow us to send messages to different, sort of, "queues" via this driver. What I mean is, you could envision something like this:

```
framework:
    messenger:
        transports:
            # Use the enqueue transport
            normal: 'doctrine://default?queue=normal'
            important: 'doctrine://default?queue=extra_important'

        routing:
            # Route your messages to the transports
            'App\Message\SmsNotification': normal
            'App\Message\EmailNotification': important
```

I think whether or not we support this is subjective. But, it seems reasonable to allow the *one* table in the database for tracking messages have multiple "queues" that can be handled independently of each other.

Reply:
@weaverryan I like the idea of the `available_at` field. 

For the `handling_at` I've already created a column named `delivered_at`. IMHO `delivered_at` is more telling than `handling_at`.

I see what you want for the multiple transport. Actually it's possible with the `table` option as documented in the doc PR : https://github.com/symfony/symfony-docs/pull/10616/files#diff-4eccd2af1652149bd44d2668873dedf8R51